### PR TITLE
[cpp-package] add lr scheduler

### PR DESCRIPTION
**Pull request by CNevd**
1. add lr scheduler to optimizer
2. move learning rate and weight decay to optimizer
3. fix examples
@lx75249

### CODE REVIEW

**Comment by CNevd** [comment]
@lx75249 have time to review this?

**Comment by CNevd** [comment]
@piiswrong

File: cpp-package/include/mxnet-cpp/executor.h [hunk 1]
```diff
@@ -79,18 +79,6 @@ class Executor {
...
-  * \brief update the arguments with given learning rate and optimizer
-  * \param opt the pointer to the optimizer
-  * \param lr learning rate
-  * \param wd weight decay
-  * \param arg_update_begin begin index of the arguments to be updated, it
-  * starts after the input data by default
-  * \param arg_update_end end index of the arguments to be updated, it ends
-  * before the label data by default
-  */
-  void UpdateAll(Optimizer *opt, float lr, float wd, int arg_update_begin = 1,
```
Comment (lx75249) [inline cpp-package/include/mxnet-cpp/executor.h]: Why remove this?

File: cpp-package/example/mlp_gpu.cpp [hunk 2]
```diff
@@ -70,12 +70,24 @@ int main(int argc, char** argv) {
 
   // Create sgd optimizer
   Optimizer* opt = OptimizerRegistry::Find("sgd");
-  opt->SetParam("rescale_grad", 1.0/batch_size);
+  opt->SetParam("rescale_grad", 1.0/batch_size)
+     ->SetParam("lr", learning_rate)
+     ->SetParam("wd", weight_decay);
+  std::unique_ptr<LRScheduler> lr_sch(new FactorScheduler(5000, 0.1));
+  opt->SetLRScheduler(std::move(lr_sch));
```
Comment (lx75249) [inline cpp-package/example/mlp_gpu.cpp]: please add the scheduler part to test_score.cpp, for integration test.

File: cpp-package/include/mxnet-cpp/optimizer.hpp [hunk 3]
```diff
@@ -100,6 +94,18 @@ inline unsigned Optimizer::UpdateCount_(int index) {
   return new_count;
 }
 
+inline float Optimizer::GetLR_(int index) {
+  float lr = std::stof(params_["lr"]);
```
Comment (lx75249) [inline cpp-package/include/mxnet-cpp/optimizer.hpp]: return directly to avoid extra string->float casting?

File: cpp-package/include/mxnet-cpp/executor.h (hunk 1 above)
Comment (CNevd) [inline cpp-package/include/mxnet-cpp/executor.h]: So we can control which data to update and put update to module later

File: cpp-package/include/mxnet-cpp/optimizer.hpp [hunk 4]
```diff
@@ -94,15 +94,15 @@ inline unsigned Optimizer::UpdateCount_(int index) {
   return new_count;
 }
 
-inline float Optimizer::GetLR_(int index) {
+inline std::string Optimizer::GetLR_(int index) {
```
Comment (lx75249) [inline cpp-package/include/mxnet-cpp/optimizer.hpp]: i meant.. return `stof(params_["lr"])` instead of `lr=stof(params_["lr"])`, so that when `lrScheduler_ != nullptr`, `stof(params_["lr"])` won't be evaluated.

Comment (CNevd) [inline cpp-package/include/mxnet-cpp/optimizer.hpp]: ok then, done

**Comment by CNevd** [comment]
@lx75249 done

**Comment by lx75249** [comment]
LGTM @piiswrong

**Comment by CNevd** [comment]
@piiswrong can we merge this?
